Homework_1/src/Homework_1.jl

```diff
@@ -0,0 +1,625 @@
+    
+Multiply the Givens rotation matrix `Q` by a vector `x`.
+Returns a new vector containing the result of the multiplication.
+"""
+function Base.:*(Q::Givens, x::Vector)
+    y = copy(x)
+    for (c, s, i, j) in reverse(Q.rotations)
+        yi = y[i]
+        yj = y[j]
+        y[i] = c * yi - s * yj
+        y[j] = s * yi + c * yj
+    end
+    return y
+end
+
+"""
+    *(Q::Givens, A::Matrix)
+
+Multiply the Givens rotation matrix `Q` by a matrix `A`.
+Returns a new matrix containing the result of the multiplication.
+"""
+function Base.:*(Q::Givens, A::Matrix)
+    B = copy(A)
+    for (c, s, i, j) in reverse(Q.rotations)
+        for col in 1:size(B, 2)
+            a = B[i, col]
+            b = B[j, col]
+            B[i, col] = c * a - s * b
+            B[j, col] = s * a + c * b
+        end
+    end
+    return B
+end
+
+"""
+    *(A::Matrix, Q::Givens)
+
+Multiply a matrix `A` by the Givens rotation matrix `Q`.
+Returns a new matrix containing the result of the multiplication.
+"""
+function Base.:*(A::Matrix, Q::Givens)
+    B = copy(A)
+    for (c, s, i, j) in Q.rotations
+        for col in 1:size(B, 1)
+            a = B[col, i]
+            b = B[col, j]
+            B[col, i] = c * a + s * b
+            B[col, j] = -s * a + c * b
+        end
+    end
+    return B
+end
+
+"""
+    isapprox(G1::Givens, G2::Givens)
+
+Check if two Givens rotation matrices `G1` and `G2` are approximately equal.
+Returns `true` if they are approximately equal, `false` otherwise.
+"""
+function isapprox(G1::Givens, G2::Givens; atol=1e-10, rtol=1e-6)
+    if length(G1.rotations) != length(G2.rotations)
+        return false
+    end
+    for (r1, r2) in zip(G1.rotations, G2.rotations)
+        if !isapprox(r1[1], r2[1]; atol=atol, rtol=rtol) || !isapprox(r1[2], r2[2]; atol=atol, rtol=rtol)
+            return false
+        end
+        # Indices i and j in Givens rotations can be swapped
+        if r1[3] == r2[3] && r1[4] == r2[4]
+            continue
+        elseif r1[3] == r2[4] && r1[4] == r2[3]
+            continue
+        else
+            return false
+        end
+    end
+    return true
+end
+
+####################################################################
+
+"""
+    *(Q::Givens, R::ZgornjeTridiag)
+
+Multiply the Givens rotation matrix `Q` by the upper triangular matrix `R`.
+Returns a Tridiag matrix if product of Givens and ZgornjeTridiag results in a tridiagonal matrix,
+    otherwise returns a full matrix. 
+"""
+function Base.:*(Q::Givens, R::ZgornjeTridiag)
+    n = length(R.gd)
+    B = [R[i, j] for i in 1:n, j in 1:n]
+    A = Q * B
+    # Check if the result is tridiagonal
+    eps = 1e-12
+    is_tridiagonal = true
+    for i in 1:n
+        for j in 1:n
+            if abs(i - j) > 1 && abs(A[i, j]) > eps
+                is_tridiagonal = false
+                break
+            end
+        end
+        if !is_tridiagonal
+            break
+        end
+    end
+    if is_tridiagonal
```

Comment:
If statemants like this represent a code smell. They are usually a hint, that we need to refine our types. They can be replaced with polymorphism. See suggestions above about the `Givens` type.